graphql-server/pkg/queries/proposal.go

```diff
@@ -25,30 +25,64 @@ type ProposalQuery struct {
 	session *bun.DB
 
 	scopedProposalStatus models.ProposalStatus
-	scopedRelatedAddress string
+	scopedAddressFilter  *models.ProposalAddressFilter
 }
 
 func NewProposalQuery(ctx context.Context, config config.Config, session *bun.DB) IProposalQuery {
 	return &ProposalQuery{ctx: ctx, config: config, session: session}
 }
 
+// generate new select query with filters applied.
+// when both status and address filters are applied, results satisfying both filters are returned.
+// when multiple sub-filters in address filter (ie. isDepositor, IsSubmitter, IsVoter), results satisfying
+// any one of the sub-filters are returned
 func (q *ProposalQuery) NewQuery() *bun.SelectQuery {
 	query := q.session.NewSelect().Model((*models.Proposal)(nil))
 
 	if q.scopedProposalStatus != "" {
 		query = query.Where("status = ?", q.scopedProposalStatus)
 	}
 
-	if q.scopedRelatedAddress != "" {
-
-		relatedDeposits := q.session.NewSelect().Model((*models.ProposalDeposit)(nil)).Column("proposal_id").Where("depositor_address = ?", q.scopedRelatedAddress)
-		relatedVotes := q.session.NewSelect().Model((*models.ProposalVote)(nil)).Column("proposal_id").Where("voter_address = ?", q.scopedRelatedAddress)
-
-		relatedProposals := relatedDeposits.Union(relatedVotes)
-
-		query = query.
-			WhereOr("id IN (?)", relatedProposals).
-			WhereOr("proposal.proposer_address = ?", q.scopedRelatedAddress)
+	if q.scopedAddressFilter != nil {
+		relatedDeposits := q.session.NewSelect().
+			Model((*models.ProposalDeposit)(nil)).
+			Column("proposal_id").
+			Where("depositor_address = ?", q.scopedAddressFilter.Address)
+
+		relatedVotes := q.session.NewSelect().
+			Model((*models.ProposalVote)(nil)).
+			Column("proposal_id").
+			Where("voter_address = ?", q.scopedAddressFilter.Address)
+
+		isFirstAddressFilter := true
```

Comment:
what does FirstAddress mean?

Reply:
`isFirstAddressFilter` check if the address filter (ie. isDepositor, isVoter or isSubmitter) being added to the query object is the first filter being added. If it is it needs to be a `Where` rather then `WhereOr`.

Reply:
it looks like they are interchangeable based on the implementation 🤔 
https://github.com/uptrace/bun/blob/e816b1a7bf524f0b9cd25579cf86e0e35bb53a02/query_select.go#L131-L139

I know they use `.Where()` + `.WhereOr()` as example in their official docs but does using only `WhereOr` produce a wrong query? bun might be unfriendly in some cases but not this unfriendly :sosad:

Reply:
I am not sure how the implementation would work if only `WhereOr` is used? 

If only use `WhereOr`, in the cases where both status and address filters are applied, it would result in SQL as below, making the 2 filters applied in union, rather than their intersection.

```sql
SELECT * 
WHERE 
    status = [desired status] OR
    voter_address = [desired address] OR depositor_address = [desired address]
```

but we want 

```sql
SELECT * 
WHERE 
    status = [desired status] AND
    voter_address = [desired address] OR depositor_address = [desired address]
```

Reply:
https://bun.uptrace.dev/guide/query-where.html#grouping

does this help you?